server/clip_server/model/transformer.py

```diff
@@ -0,0 +1,315 @@
+# Originally from https://github.com/mlfoundations/open_clip.
+#
+# Copyright (c) 2012-2021 Gabriel Ilharco, Mitchell Wortsman,
+# Nicholas Carlini, Rohan Taori, Achal Dave, Vaishaal Shankar,
+# John Miller, Hongseok Namkoong, Hannaneh Hajishirzi, Ali Farhadi,
+# Ludwig Schmidt
+
+
+from collections import OrderedDict
+from typing import Callable, Optional
+
+import torch
+from torch import nn
+from torch.nn import functional as F
+from torch.utils.checkpoint import checkpoint
+
+from open_clip.utils import to_2tuple
+from open_clip.transformer import LayerNorm
+
+# Use flash attention
+try:
+    from clip_server.model.flash_attention import MultiheadAttention
+
+    FLASH_ATTENTION_AVAILABLE = True
+except:
+    FLASH_ATTENTION_AVAILABLE = False
+
+
+class ResidualAttentionBlock(nn.Module):
+    def __init__(
+        self,
+        d_model: int,
+        n_head: int,
+        mlp_ratio: float = 4.0,
+        act_layer: Callable = nn.GELU,
+        norm_layer: Callable = LayerNorm,
+        scale_attn: bool = False,
+        scale_fc: bool = False,
+        cast_dtype=torch.float32,
+    ):
+        super().__init__()
+        head_dim = d_model // n_head
+        self.flash_attention = head_dim % 8 == 0 and head_dim <= 128
+
+        self.ln_1 = norm_layer(d_model)
+        self.attn = (
+            MultiheadAttention(d_model, n_head)
+            if FLASH_ATTENTION_AVAILABLE
+            and torch.cuda.is_available()
+            and cast_dtype in (torch.float16, torch.bfloat16)
+            and self.flash_attention
+            else nn.MultiheadAttention(d_model, n_head)
+        )
+        self.ln_attn = LayerNorm(d_model) if scale_attn else nn.Identity()
+
+        self.ln_2 = norm_layer(d_model)
+        mlp_width = int(d_model * mlp_ratio)
+        self.mlp = nn.Sequential(
+            OrderedDict(
+                [
+                    ("c_fc", nn.Linear(d_model, mlp_width)),
+                    ('ln', norm_layer(mlp_width) if scale_fc else nn.Identity()),
+                    ("gelu", act_layer()),
+                    ("c_proj", nn.Linear(mlp_width, d_model)),
+                ]
+            )
+        )
+
+    def attention(self, x: torch.Tensor, attn_mask: Optional[torch.Tensor] = None):
+        attn_mask = attn_mask.to(x.dtype) if attn_mask is not None else None
+        return self.attn(x, x, x, need_weights=False, attn_mask=attn_mask)[0]
+
+    def forward(self, x: torch.Tensor, attn_mask: Optional[torch.Tensor] = None):
```

Comment:
Do we need cast the dtype based on `cast_dtype` in `forward`?

Reply:
`TextTransformer.forward` already casted